### ISSUE TITLE
test_processingFailedDisplayTracebackHandlesUnicode fails on Windows with Python 3.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8778]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8778|
|Type|enhancement|
|Created|2016-08-21 17:04:40Z|
|Branch|https://github.com/twisted/twisted/tree/8778-rodrigc-web-py3|

If I do this:

```
python -m twisted.trial twisted.web.test.test_web.RequestTests.test_processingFailedDisplayTracebackHandlesUnicode 
```

I get:

```
[FAIL]
Traceback (most recent call last):
  File "C:\Users\Craig\twisted3\src\twisted\web\test\test_web.py", line 629, in test_processingFailedDisp
layTracebackHandlesUnicode
    self.assertEqual(1, len(self.flushLoggedErrors()))
  File "C:\Users\Craig\twisted3\src\twisted\trial\_synctest.py", line 425, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "C:\Python35\Lib\unittest\case.py", line 820, in assertEqual
    assertion_func(first, second, msg=msg)
  File "C:\Python35\Lib\unittest\case.py", line 813, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 1 != 2

twisted.web.test.test_web.RequestTests.test_processingFailedDisplayTracebackHandlesUnicode
-------------------------------------------------------------------------------
```

It looks like on Windows with Python 3, this test is generating two
Failure objects:

```
 b"[Failure instance: Traceback (failure with no frames): <class 'Exception'>: \xe2\x98\x83\n]"

```
```
1 b"[Failure instance: Traceback: <class 'UnicodeEncodeError'>: 'charmap' codec can't encode character '\\u2603' in position 113: character maps to <undefined>\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\python\\log.py:132:err\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\python\\threadable.py:53:sync\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\python\\log.py:282:msg\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\logger\\_legacy.py:154:publishToNewObserver\n
--- <exception caught here> ---\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\logger\\_observer.py:131:__call__\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\logger\\_legacy.py:93:__c
all__\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\python\\log.py:557:emit\n
C:\\Users\\Craig\\twisted3\\src\\twisted\\python\\util.py:715:untilConcludes\n
C:\\Python35\\Lib\\encodings\\cp1252.py:19:encode\n]"
```

<details><summary>Searchable metadata</summary>

```
trac-id__8778 8778
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8778_rodrigc_web_py3 8778-rodrigc-web-py3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1471799080575196 1471799080575196
changetime__1471820961861987 1471820961861987
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

On Python 2, the same UnicodeError is thrown, but only one Failure object is created:

```
'[Failure instance: Traceback (failure with no frames): <type \'exceptions.Exception\'>: <Exception instance at 0x4b11ac8 with str error:\n Traceback (most recent call last):\n
File "twisted\\python\\reflect.py", line 440, in safe_str\n
    return str(o)\nUnicodeEncodeError: \'ascii\' codec can\'t encode character
u\'\\u2603\' in position 0: ordinal not in range(128)\n>\n]')
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

The source of the UnicodeError is this in twisted/trial/runner.py:

```
 def __init__(self, reporterFactory,
              mode=None,
              logfile='test.log',
              stream=sys.stdout,
              profile=False,
              tracebackFormat='default',
              realTimeErrors=False,
              uncleanWarnings=False,
              workingDirectory=None,
              forceGarbageCollection=Fa
lse,
              debugger=None,
              exitFirst=False):
```

The default value of stream is sys.stdout.
On Windows, sys.stdout.encoding is equal to cp437.
So when the code tries to log the Failure using the cp437 codec, it fails because of
unicode characters which are not handled by cp437.

On other platforms, the value of sys.stdout.encoding is UTF-8.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/pull/517

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|meejah set owner to meejah|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|meejah commented|
|-|-|

Looks good to me. +1

**Comment by twisted-trac** [comment]
|<img alt="meejah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|meejah set owner to @rodrigc|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset 1698513f582bc65504a33eef9944b64d10d1de0c
```
#!CommitTicketReference repository="" revision="1698513f582bc65504a33eef9944b64d10d1de0c"
Merge pull request #517 from twisted/8778-rodrigc-web-py3

Author: rodrigc
Reviewer: meejah
Fixes: #8778

Fix test_web on Windows with Python 3.5
```
